mzLib/Readers/timsTOF/TimsTofFileReader.cs

```diff
@@ -0,0 +1,566 @@
+        }
+
+        internal void OpenSqlConnection()
+        {
+            _sqlConnection = new SQLiteConnection("Data Source=" +
+                Path.Combine(FilePath, "analysis.tdf") +
+                "; Version=3");
+            try
+            {
+                _sqlConnection.Open();
+            }
+            catch (Exception e)
+            {
+                throw new MzLibException("Error opening the .tdf file: " + e.Message);
+            }
+        }
+
+        internal void OpenBinaryFileConnection()
+        {
+            byte[] binaryFileBytePath = BrukerFileReader.ConvertStringToUTF8ByteArray(FilePath);
+            _fileHandle = tims_open(binaryFileBytePath, 0);
+            if (_fileHandle == null || _fileHandle == 0)
+                throw new MzLibException("Could not open the analysis.tdf_bin file");
+        }
+
+        public override void CloseDynamicConnection()
+        {
+            _sqlConnection?.Close();
+            _sqlConnection?.Dispose();
+            if (_fileHandle != null)
+            {
+                tims_close((UInt64)_fileHandle);
+                _fileHandle = null;
+            }   
+        }
+
+        /// <summary>
+        /// WARNING! This method reads in the entire data file before
+        /// returning the requested scan! It is recommended to call the 
+        /// GetScanFromPrecursorAndFrameIdFromDynamicConnection()
+        /// </summary>
+        public override MsDataScan GetOneBasedScanFromDynamicConnection(int oneBasedScanNumber, IFilteringParams filterParams = null)
+        {
+            if(oneBasedScanNumber <= 0)
+                throw new IndexOutOfRangeException("Invalid one-based index given when accessing data scans. Index: " + oneBasedScanNumber);    
+            if (Scans != null && Scans.Length >= oneBasedScanNumber && Scans[oneBasedScanNumber - 1] != null)
+                return Scans[oneBasedScanNumber - 1];
+
+            LoadAllStaticData(filteringParams: (FilteringParams)filterParams);
+            if (oneBasedScanNumber > Scans.Length)
+                throw new IndexOutOfRangeException("Invalid one-based index given when accessing data scans. Index: " + oneBasedScanNumber);
+            return Scans[oneBasedScanNumber - 1];
+        }
+
+        /// <summary>
+        /// Returns a TimsDataScan with the specified frame and precursor id
+        /// WARNING! The returned data scan will have a OneBasedScanNumber of -1
+        /// </summary>
+        /// <param name="precursorId"></param>
+        /// <param name="frameId"></param>
+        /// <param name="filteringParams"></param>
+        /// <returns></returns>
+        /// <exception cref="MzLibException"></exception>
+        /// <exception cref="ArgumentException"></exception>
+        /// <exception cref="NotImplementedException"></exception>
+        public TimsDataScan? GetScanFromPrecursorAndFrameIdFromDynamicConnection(int precursorId, int frameId, IFilteringParams filteringParams = null)
```

Comment:
You throw on null recordForPrecursor, does this need to be nullable in that case?